src/pipeline.rs

```diff
@@ -0,0 +1,174 @@
+use windows::Win32::Foundation::{HWND, LPARAM, WPARAM};
+use windows::Win32::Graphics::Direct3D12::ID3D12Resource;
+#[cfg(target_arch = "x86")]
+use windows::Win32::UI::WindowsAndMessaging::SetWindowLongA;
+use windows::Win32::UI::WindowsAndMessaging::{SetWindowLongPtrA, GWLP_WNDPROC};
+
+use crate::renderer::input::{imgui_wnd_proc_impl, WndProcType};
+use crate::renderer::RenderEngine;
+use crate::{util, ImguiRenderLoop};
+
+type RenderLoop = Box<dyn ImguiRenderLoop + Send + Sync>;
+
+pub struct PipelineMessage(pub HWND, pub u32, pub WPARAM, pub LPARAM);
+
+pub struct PipelineSharedState {
+    pub should_block_events: AtomicBool,
+    pub wnd_proc: WndProcType,
+    pub tx: Sender<PipelineMessage>,
+}
+
+pub struct Pipeline<T> {
+    hwnd: HWND,
+    compositor: T,
+    ctx: Context,
+    engine: RenderEngine,
+    render_loop: RenderLoop,
+    rx: Receiver<PipelineMessage>,
+    shared_state: Arc<PipelineSharedState>,
+}
+
+impl<T> Pipeline<T> {
+    pub fn new(
+        hwnd: HWND,
+        wnd_proc: WndProcType,
+        compositor: T,
+        mut render_loop: RenderLoop,
+    ) -> std::result::Result<(Self, Arc<PipelineSharedState>), (Error, RenderLoop)> {
+        let (width, height) = util::win_size(hwnd);
+
+        let mut ctx = Context::create();
+        ctx.io_mut().display_size = [width as f32, height as f32];
+
+        let mut engine = match RenderEngine::new(&mut ctx, width as u32, height as u32) {
+            Ok(engine) => engine,
+            Err(e) => return Err((e, render_loop)),
+        };
+
+        render_loop.initialize(&mut ctx, &mut engine);
+
+        #[cfg(any(target_arch = "aarch64", target_arch = "x86_64"))]
+        let wnd_proc = unsafe {
+            mem::transmute(SetWindowLongPtrA(hwnd, GWLP_WNDPROC, wnd_proc as usize as isize))
+        };
+
+        // TODO is this necessary? SetWindowLongPtrA should already decay to
+        // SetWindowLongA
+        #[cfg(target_arch = "x86")]
+        let wnd_proc =
+            unsafe { mem::transmute(SetWindowLongA(hwnd, GWLP_WNDPROC, wnd_proc as usize as i32)) };
+
+        let (tx, rx) = mpsc::channel();
+        let shared_state = Arc::new(PipelineSharedState {
+            should_block_events: AtomicBool::new(false),
+            wnd_proc,
+            tx,
+        });
+
+        Ok((
+            Self {
+                hwnd,
+                compositor,
+                ctx,
+                engine,
+                render_loop,
+                rx,
+                shared_state: Arc::clone(&shared_state),
+            },
+            shared_state,
+        ))
+    }
+
+    pub fn render(&mut self) -> Result<ID3D12Resource> {
+        // TODO find a better alternative than allocating each frame
+        let message_queue = self.rx.try_iter().collect::<Vec<_>>();
+
+        message_queue.into_iter().for_each(|PipelineMessage(hwnd, umsg, wparam, lparam)| {
+            imgui_wnd_proc_impl(hwnd, umsg, wparam, lparam, self);
+        });
```

Comment:
Pass directly to for_each?

```suggestion
        self.rx.try_iter().for_each(|PipelineMessage(hwnd, umsg, wparam, lparam)| {
            imgui_wnd_proc_impl(hwnd, umsg, wparam, lparam, self);
        });
```

Reply:
Unfortunately not: `self.rx.try_iter()` already mutably borrows `self`, and `imgui_wnd_proc_impl` would also have to mutably borrow `self` for the same lifetime which is not possible.

Similarly, allocating a vector in `Pipeline` and draining/extending it there with the contents of `try_iter` is also not going to work for the same reason, barring using `RefCell` which has overhead anyway -- maybe less so than allocating a vector outright, but at the cost of making the code look a lot uglier.

Reply:
I have concocted a solution for this:

- Store an allocated `Vec` in a `OnceCell`.
- Pull it out of the `OnceCell` before this section and use it as a backing storage for `rx.try_iter()`.
- Drain it immediately (preserving the capacity) and pass the contents to the wnd proc implementation.
- Store the drained `Vec` back in the `OnceCell`.

This is a bit ugly, but works around the lifetime issues with little to no performance overhead -- `OnceCell::take` and `OnceCell::set` are simple moves, memory is only ever reallocated when needed, a generous buffer could be pre-allocated, and even a large vector won't be that much of a memory waste.